### PR TITLE
Adds User.all()

### PR DESCRIPTION
**Pull request by cmelchior**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ### Enhancement
 
 * `Realm.compactRealm()` works for encrypted Realms.
+*  Added `User.all()` that returnal all known Realm Object Server users.
```
Comment (Zhuinden) [inline CHANGELOG.md]: returnal?

Comment (spl) [inline CHANGELOG.md]: ~~returnal~~ returns

Comment (spl) [inline CHANGELOG.md]: Double shot! 👊 👊

Comment (cmelchior) [inline CHANGELOG.md]: I was just testing if you were awake 😛

Comment (spl) [inline CHANGELOG.md]: Too bad GitHub doesn't show timestamps with a resolution smaller than a minute so we could see who won that (comment) race.

File: realm/realm-library/src/objectServer/java/io/realm/User.java [hunk 2]
```diff
@@ -80,6 +82,24 @@ public static User currentUser() {
     }
 
     /**
+     * Returns all valid users known by this device.
+     * A user is invalidated when he/she logs out or the user's access token expire.
+     *
+     * @return a list of all known users
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/User.java]: `.` at the end.

File: realm/realm-library/src/objectServer/java/io/realm/User.java [hunk 3]
```diff
@@ -80,6 +82,24 @@ public static User currentUser() {
     }
 
     /**
+     * Returns all valid users known by this device.
+     * A user is invalidated when he/she logs out or the user's access token expire.
```
Comment (beeender) [inline realm/realm-library/src/objectServer/java/io/realm/User.java]: `expire` -> `expires`

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/UserTests.java [hunk 4]
```diff
@@ -68,6 +76,27 @@ public void currentUser_returnsNullIfUserExpired() {
...
+    public void all_validUsers() {
+        // Add 1 expired user and 1 valid user to the user store
+        UserStore userStore = new SharedPrefsUserStore(InstrumentationRegistry.getContext());
+        SyncManager.setUserStore(userStore);
+        userStore.put(UserStore.CURRENT_USER_KEY, SyncTestUtils.createTestUser(Long.MIN_VALUE));
+        userStore.put(UserStore.CURRENT_USER_KEY, SyncTestUtils.createTestUser(Long.MAX_VALUE));
+
+        Collection<User> users = User.all();
+        assertEquals(1, users.size());
+
```
Comment (kneth) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/UserTests.java]: Remove empty line

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/UserTests.java [hunk 5]
```diff
@@ -68,6 +76,27 @@ public void currentUser_returnsNullIfUserExpired() {
...
+    @Test
+    public void all_validUsers() {
+        // Add 1 expired user and 1 valid user to the user store
+        UserStore userStore = new SharedPrefsUserStore(InstrumentationRegistry.getContext());
+        SyncManager.setUserStore(userStore);
+        userStore.put(UserStore.CURRENT_USER_KEY, SyncTestUtils.createTestUser(Long.MIN_VALUE));
+        userStore.put(UserStore.CURRENT_USER_KEY, SyncTestUtils.createTestUser(Long.MAX_VALUE));
+
+        Collection<User> users = User.all();
+        assertEquals(1, users.size());
```
Comment (kneth) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/UserTests.java]: Please test if the valid user is returned.

File: realm/realm-library/src/androidTestObjectServer/java/io/realm/UserTests.java [hunk 6]
```diff
@@ -68,6 +76,27 @@ public void currentUser_returnsNullIfUserExpired() {
...
+    public void all_empty() {
+        Collection<User> users = User.all();
+        assertTrue(users.isEmpty());
+    }
+
+    // `all()` returns only valid users. Invalid users are filtered.
+    @Test
+    public void all_validUsers() {
+        // Add 1 expired user and 1 valid user to the user store
+        UserStore userStore = new SharedPrefsUserStore(InstrumentationRegistry.getContext());
```
Comment (cmelchior) [inline realm/realm-library/src/androidTestObjectServer/java/io/realm/UserTests.java]: The SharedPreference is a Android concept which require the Context. The `UserStore` is our general interface that enable us to swap implementations for the relevant platforms. The `UserStore` do not have any knowledege about Context

**Pull request merged**